src/librustc_typeck/check/demand.rs

```diff
@@ -47,8 +53,48 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 assert!(obligations.is_empty());
             },
             Err(e) => {
-                self.report_mismatched_types(origin, expected, actual, e);
+                self.report_mismatched_types(origin, expected, actual, e).emit();
+            }
+        }
+    }
+
+    fn check_ref(&self, expr: &hir::Expr, checked_ty: Ty<'tcx>,
+                 expected: Ty<'tcx>) -> Option<String> {
+        match (&checked_ty.sty, &expected.sty) {
+            (&ty::TyRef(_, x_mutability), &ty::TyRef(_, y_mutability)) => {
```

Comment:
Strange... Anywhere else this is the opposite... Well whatever, I don't mind so I'll just switch.